src/Orleans/Utils/IKeyedServiceCollection.cs

```diff
@@ -10,7 +12,40 @@ namespace Orleans.Runtime
     public interface IKeyedServiceCollection<in TKey, out TService>
         where TService : class
     {
-        TService GetService(TKey key);
+        TService GetService(IServiceProvider services, TKey key);
+    }
+
+    public interface IKeyedService<TKey, out TService> : IEquatable<TKey>
+    {
+        TService GetService(IServiceProvider services);
+    }
+
+    public class KeyedService<TKey, TService, TInstance> : IKeyedService<TKey, TService>
+            where TInstance : TService
+    {
+        private readonly TKey key;
+
+        public KeyedService(TKey key)
+        {
+            this.key = key;
+        }
+
+        public TService GetService(IServiceProvider services) => services.GetService<TInstance>();
```

Comment:
I'll add more comments, but this keyed service is simply a mapping between a name and a type.  If the type is registered as being transient (as are the only cases currently supported by extension functions) the named type will be transient.  If the type is registered as a singleton or scoped, those properties will also apply.

In AddTransientKeyedService we add the instance type as transient but the mapping from name to type as a singleton.  So the named type will be transient, as would be expected from the name of the extension function.

    collection.TryAddTransient<TInstance>(); 
    collection.AddSingleton<IKeyedService<TKey, TService>>(sp => new KeyedService<TKey, TService, TInstance>(key)); 

However, we could have an AddScopedKeyedService extension function that does the following

    collection.TryAddScoped<TInstance>(); 
    collection.AddSingleton<IKeyedService<TKey, TService>>(sp => new KeyedService<TKey, TService, TInstance>(key)); 

Which would created a scoped service by name.  Same can be done for singletons.